### PR TITLE
Periodically update the slices in the dashboard

### PR DESCRIPTION
**Pull request by x4base**

### CODE REVIEW

**Comment by philippfrenzel** [comment]
what happens if you go offline? Or you don't wanna have a refresh every 60 seconds?

**Comment by x4base** [comment]
It would be ideal if the users can freely decide the refresh interval ( or never refresh) for each slice. But in the short term, maybe it we can just make it configurable in a config file? I just don't have any idea where is the better place for javascript configuration or whether is already any. Do you have any idea? Thanks

**Comment by x4base** [comment]
@philippfrenzel  I added the UI so the user can choose the refresh interval now

**Comment by mistercrunch** [comment]
I don't think the callback approach is necessary, the dashboard controller can manage this on its own.

**Comment by x4base** [comment]
@mistercrunch In my use case, some queries are heavy. For example, if the user wants the dashboard to refresh every 10 seconds, but the query actually takes 30s, without the callback, there can be too many http requests pending in the browser. Maybe the user should choose a 1-minute interval, but maybe the user just doesn't know they are making the server load heavy. So I think a callback is natural for an async operation like this. Does it make sense to you?

**Comment by mistercrunch** [comment]
The frequency of updates should be much lower than the duration of the update. Regardless if any post refresh time-stamping it should be done in the `Slice.done` method. 

I was also thinking there should be a 10-ish% randomness on the delay so that not all widgets refresh at the same time. The parameter should be set on a per-dashboard basis.

**Comment by x4base** [comment]
@mistercrunch I implemented it and added an upper bound of 5 seconds. Please check again

**Comment by mistercrunch** [comment]
Nice! This is starting to look pretty solid. I'll do a more thorough round of review now.

**Comment by mistercrunch** [comment]
One thing we need to address is how the loading image pushes everything down while it's showing, it looks jittery. Maybe it should render as an overlay of some kind.

File: caravel/assets/javascripts/dashboard.js [hunk 1]
```diff
@@ -57,6 +58,36 @@ var Dashboard = function (dashboardData) {
...
+    },
+    startPeriodicRender: function (interval) {
+      this.stopPeriodicRender();
+      var dash = this;
+      var maxRandomDelay = Math.min(interval * 0.1, 5000);
+      var refreshAll = function () {
+        dash.slices.forEach(function (slice) {
+          setTimeout(function () {
+                slice.render(true);
+              },
```
Comment (mistercrunch) [inline caravel/assets/javascripts/dashboard.js]: indent is off here

**Comment by mistercrunch** [comment]
I think apart from the loading jitter it looks great and I'll merge as soon as you address it!

**Comment by x4base** [comment]
It's done. Please check again

**Comment by mistercrunch** [comment]
img

**Comment by gbrian** [comment]
+1
What do you think about adding granularity & refresh control like Grafana:

image

Thanks

PS:
Typo on code:
<i class="fa fa-clock-o" data-toggle="tooltip" title="**Edit the dashboard's CSS**"></i>

**Comment by x4base** [comment]
Thank you for pointing it out. I opened a pull request for that ( #386 ).
Also, are you mainly talking about the floating dialog like that, or the Quick ranges? Or a refresh interval selection box in the slice editing page?

**Comment by gbrian** [comment]
Hi, 
I mean both. Refresh is already added, I'll suggest having the availability to overwrite granularity as well.

Will be great to have `Rows` basically multiple dashboards so we can define refresh/granularity for different `parts` of the dashboard.

Think on this:

 - First row: Daily bookings KPI (refresehd each 5 min)
 - Second row: YTD / MTD / WTD KPI (refreshed each 1h) 
 - Third row: Hostoric data (refreshed each 2 hours)

First and Second rows share same dashboard but with different granularity/refresh time.

**Pull request merged**